OSX Build fixes

rvt opened the pull request:
Fixes to build on OSX work again

Tested on OSX 10.14 & 10.15

**Build Instructions with brew:**

**Build opentrack-depends**
```
cd ~/Developer
git clone https://github.com/opentrack/opentrack-depends
cd opentrack-depends

# Example build aruco
cd aruco
mkdir build
cd build
cmake -DCMAKE_BUILD_TYPE=Release ..
make
```

**Build opentrack**
```
brew install eigen opencv cmake qt5 libusb create-dmg imagemagick

export Qt5_DIR="/usr/local/opt/qt5"
export PATH="/usr/local/opt/qt5/bin:$PATH"

cd ~/Developer
git clone https://github.com/opentrack/opentrack
cd opentrack
mkdir build
cd build
cmake ..
make -j 4 install

# Find your DMG in the build directory
```

For debug builds use `cmake -DCMAKE_BUILD_TYPE=Debug ..`

**Add opentrack depends to your build:**
Create a file called `sdk-paths-<USERNAME>@Clang-Darwin.cmake` with the following contents:
```
function(set_sdk var path)
    set(${var} "$ENV{HOME}/Developer/opentrack-depends/${path}" CACHE INTERNAL "" FORCE)
endfunction()

set_sdk(SDK_ARUCO_LIBPATH "aruco/build/src/libaruco.a")
```

After that re-compile opentrack, make sure you you don't get something like `-- User settings file 'sdk-paths-<USERNAME>@Clang-Darwin.cmake' doesn't exist`

**Note: on App Nap**
To keep the application running in the background run tthe following from the terminal:
`defaults write com.github.opentrack NSAppSleepDisabled -bool YES`

Work is underway to set this from the application.



![screenshot](https://i.imgur.com/fuXfngK.png)

sthalik commented:
> Need to create a package/installer whatever works best

Running `make install` should create an `.app` package.

rvt commented:
> > Need to create a package/installer whatever works best
> 
> Running `make install` should create an `.app` package.

I was actually thinking for apple in the line of CPack/CPackDMG Quickly glancing over the directory that cmake install created it doesn't look like this can be distributed as-is and needs to have the proper libraries added because it does not look like it's done yet.

rvt commented on cmake/FindEigen3.cmake:
```diff
@@ -61,19 +67,29 @@ if (EIGEN3_INCLUDE_DIR)
   endif(EIGEN3_INCLUDE_DIR)
 
-  #include(FindPackageHandleStandardArgs)
-  #find_package_handle_standard_args(Eigen3 DEFAULT_MSG EIGEN3_INCLUDE_DIR EIGEN3_VERSION_OK)
```
Not 100% sure if this will work on WIndows/Linux. Tested in OSX it was fine

sthalik commented:
One major feature missing from OSX support is hotkeys. There's no support for keyup events, and non-alphanumeric keys mostly don't work.

rvt commented:
> Running `make install` should create an `.app` package.

I noticed that the `make-app-bundle.sh` files was not set to executable and now indeed a package is created. However it's not bundled correctly.
When run on a system where qt5 was not installed you get a message like this:

```
 ./opentrack.bin
dyld: Library not loaded: /usr/local/opt/qt/lib/QtWidgets.framework/Versions/5/QtWidgets
  Referenced from: /Users/QE25ID/Applications/opentrack.app/Contents/MacOS/./opentrack.bin
  Reason: image not found
Abort trap: 6
```

I looked into the make files but they look very alien to me. I can give it a shot though.... I still think that CPack with DragNDrop is the way to go as it as I understand fix the linkage issue and include the correct frameworks . If think the frameworks should have been included using this method I will look a bit deeper.

sthalik commented:
> When run on a system where qt5 was not installed you get a message like this:

It worked for me when using non-framework Qt. That is, I compiled it on my own and disabled framework support. You'd need qtbase, qttools, and qtserialport.

sthalik commented:
> Referenced from: /Users/QE25ID/Applications/opentrack.app/Contents/MacOS/./opentrack.bin

Alternately you can modify `install-fail-tool` to fix it.

rvt commented:
Just a quick update. With a few changes I have managed Cmake build a app with the frameworks included like Qt but still working out some issues and verifying the generate binary.

rvt commented:
quick update, couldn’t manage yet to add Qt plugins during thefixup. I found DeployQt5 but so fara it doesn’t tie in yet nicely with this way of building with CMake...

rvt commented:
I tried left and right with BundleUtilities and DeployQt5 but couldn't make that work, each time there seems to missing items so I reverted back to using good old macqtdeploy... And even that had some issues.... sigh...

However, I got a working app that was build on OSX 10.14 even when Cellar (from brew) is removed the applications boots up and finds all shared libraries within the app. Also works on a other mac (10.15)

Next up... 
We can generate a DMG with the sourcecode, doc and app added (drag and drop), or do we rather want to have this added to the bundle within the app?

After that I hope somebody within the community can give it a try..

sthalik commented:
> We can generate a DMG with the sourcecode, doc and app added (drag and drop), or do we rather want to have this added to the bundle within the app?

I'm completely unfamiliar with that. Do what you think is best.

> I tried left and right with BundleUtilities and DeployQt5 but couldn't make that work, each time there seems to missing items so I reverted back to using good old macqtdeploy... And even that had some issues.... sigh...

It works perfectly fine with a non-framework Qt build.

sthalik commented:
> quick update, couldn’t manage yet to add Qt plugins during thefixup.

That's what `qt.conf` is for.

rvt commented:
qt.conf is mow generated..

rvt commented:
> > We can generate a DMG with the sourcecode, doc and app added (drag and drop), or do we rather want to have this added to the bundle within the app?
> 
> I'm completely unfamiliar with that. Do what you think is best.

Will make the DMG that can be mounted so users can see the plane plugin and doc..
> 
> > I tried left and right with BundleUtilities and DeployQt5 but couldn't make that work, each time there seems to missing items so I reverted back to using good old macqtdeploy... And even that had some issues.... sigh...
> 
> It works perfectly fine with a non-framework Qt build.

if everything is Framework is relative easy (so they say....) in this build 60 libraries are added and fixed up. Even Brew shows that there are issues regarding the use of absolute paths in there libraries. All these are also fixed, thanks to your  `install-fail-tool` (I have slightly modified it, aswel as `make-app-bundle.sh`

rvt commented:
Small Update..

CI build can be found here : https://travis-ci.org/rvt/opentrack and I can send a PR for the travis config. For the moment I decided to leave ps3eye-driver alone due to build issues, but I currently have other priorities. I can pick that up later, I don't think it's a big change...

two more things left:

- make a Drag n Drop DMG
- I have an issue where the application start but somehow the configuration is not loaded when the application is started from finder. When started from terminal it works fine!

sthalik commented:
> I have an issue where the application start but somehow the configuration is not loaded when the application is started from finder. When started from terminal it works fine!

In `options/globals.cpp`, look for `static QString ini_directory_()`.

> make a Drag n Drop DMG

In my opinion it's not that important to make an easy installer if the project itself is in rough shape. Note the keybinding support.

rvt commented:
Drag and drop was rather simple to add using https://github.com/andreyvit/create-dmg

For the keybinding, what is the actual issue? I have some minimal experience with opentrack so I am not sure what I am missing out..
What I can do i be in a other application, press 's' (my binding for start) and it starts tracking. There is an issue that the camera does not start because I think OSX is preventing that. Not a big deal for this MVP...

If you have time (and OSX) can you try https://github.com/rvt/opentrack/releases/tag/opentrack-2.3.11-beta3-427-g4a0e2145 ?? and see what you see?

On this computer I have a hard time reproducing the issue, but when I move to DMG to my other computer (that has an upgraded OSX) I see the issue. But I don't have a debugger and all there :)
Just trying to narrow down what the issue might be..

sthalik commented:
> For the keybinding, what is the actual issue? I have some minimal experience with opentrack so I am not sure what I am missing out..

It's necessary to have access to at least 1-2 keybindings at all times. Games run at fullscreen and right now only alphanumeric keys are guaranteed to work.

Current code is based on obsolete libqxt and has an additional issue as in lack of keyup events.

> What I can do i be in a other application, press 's' (my binding for start) and it starts tracking. There is an issue that the camera does not start because I think OSX is preventing that. Not a big deal for this MVP...

Users want to be able to type in text fields when multitasking with alt-tab running. Which is why keys like insert, F11, or keypad need to work. I'm not sure how the OS eating keys factors into modifiers, e.g. will shift+F11 eat lone F11?

> If you have time (and OSX) can you try https://github.com/rvt/opentrack/releases/tag/opentrack-2.3.11-beta3-427-g4a0e2145 ?? and see what you see?

I don't have a VM ready, sorry :(

There's one big issue where the software expects a relative path, see: <<https://github.com/opentrack/opentrack/pull/1006/files/8d29288ecd592bae5b0bc2951cd44c7400e36654..96316119fd57b1d5acdfc0ff51e1766a9311f11f#diff-0307fc41e144da5bb17d09bd2e4bf597R13>>.

rvt commented:
For the configuration issue I think it's due to the fact that in OSX 10.15 the applications needs to request access to the documents folder. On my dev laptop I have zero issues, On my other Mac I see the issue. see [NSDocumentsFolderUsageDescription](https://developer.apple.com/documentation/bundleresources/information_property_list/nsdocumentsfolderusagedescription?language=objc)

I have updated Info.plist to see what gives...

Questions:
1) If you have time, can you do a review of the changes?
2) Do you know people that have a Mac or can we do a call for testers somewhere?

rvt commented:
> There's one big issue where the software expects a relative path, see: <https://github.com/opentrack/opentrack/pull/1006/files/8d29288ecd592bae5b0bc2951cd44c7400e36654..96316119fd57b1d5acdfc0ff51e1766a9311f11f#diff-0307fc41e144da5bb17d09bd2e4bf597R13>.

I don't think there is an (big) issue:

see:
```
bash-3.2$ otool -L opentrack
opentrack:
	/System/Library/Frameworks/Carbon.framework/Versions/A/Carbon (compatibility version 2.0.0, current version 162.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.0.0)
	opentrack-user-interface.dylib (compatibility version 0.0.0, current version 0.0.0)
	opentrack-migration.dylib (compatibility version 0.0.0, current version 0.0.0)
	opentrack-logic.dylib (compatibility version 0.0.0, current version 0.0.0)
	opentrack-qxt-mini.dylib (compatibility version 0.0.0, current version 0.0.0)
	opentrack-spline.dylib (compatibility version 0.0.0, current version 0.0.0)
	opentrack-pose-widget.dylib (compatibility version 0.0.0, current version 0.0.0)
	opentrack-api.dylib (compatibility version 0.0.0, current version 0.0.0)
	opentrack-options.dylib (compatibility version 0.0.0, current version 0.0.0)
	opentrack-compat.dylib (compatibility version 0.0.0, current version 0.0.0)
	@executable_path/../Frameworks/QtDBus.framework/Versions/5/QtDBus (compatibility version 5.14.0, current version 5.14.0)
	@executable_path/../Frameworks/QtWidgets.framework/Versions/5/QtWidgets (compatibility version 5.14.0, current version 5.14.0)
	@executable_path/../Frameworks/QtMultimedia.framework/Versions/5/QtMultimedia (compatibility version 5.14.0, current version 5.14.0)
	@executable_path/../Frameworks/QtGui.framework/Versions/5/QtGui (compatibility version 5.14.0, current version 5.14.0)
	@executable_path/../Frameworks/QtNetwork.framework/Versions/5/QtNetwork (compatibility version 5.14.0, current version 5.14.0)
	@executable_path/../Frameworks/QtCore.framework/Versions/5/QtCore (compatibility version 5.14.0, current version 5.14.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 800.7.0)
```

But I can see if I can make it the @executable_path. Finder will also start the application from within the MacOS directory, so I don't think there is a big issue. I had issues with this trying to have plugins and libraries separate... and setting up correct runtime linking..

Not sure if I can help at this time with libqtx due to time constrains...

sthalik commented:
> I don't think there is an (big) issue

`cmake/opentrack-hier.cmake`:

```cmake
include_guard(GLOBAL)

set(opentrack-install-rpath "")
if(APPLE)
    set(opentrack-hier-pfx ".")
    set(opentrack-hier-path "/")                        # MUST HAVE A TRAILING BACKSLASH
    set(opentrack-hier-pfx "${CMAKE_INSTALL_PREFIX}/opentrack.app/Contents/MacOS/Plugins")
```

This path is embedded in the binary, hence must be relative to the executable.

> If you have time, can you do a review of the changes?

I'll finish picking at it in a day or two.

> Do you know people that have a Mac or can we do a call for testers somewhere?

Unfortunately, no active people unlike Windows and Linux. Are you using the program? Worst case I can boot a VM but that's not for any kind of gaming.

rvt commented:
macdeployqt will fix the paths for use during assembly of the total app, as long as it can find them it can put them in the right place. 
For what I can see and understand is that opentrack-hier-pfx is only used for the Plugins. Opentrack itself will find them in /Plugins, so far this seems to work just fine. Perhaps I am confused between what opentrack-hier-str/opentrack-hier-pfx means. What do these acronym mean?
- hier
- str
- pfx 

I will take a step back and see the original config for Apple and perhaps I will have an Ahhhhhh!! light build above my head :) and adjust accordingly :)

rvt commented:
Folder permission has been fixed! It was indeed Catalina that required permissions but because opentrack was started with a bash script this didn't work well. A inter net search showed that Gimp, emacs and other applications also suffered from that.

Now opentrack correctly asks for permission.

When the DMG is openend:
<img width="801" alt="image" src="https://user-images.githubusercontent.com/720814/71544567-7ba2ac80-2981-11ea-8ff1-6b7fdd8e9d03.png">

Requesting Folder permission:
<img width="423" alt="image" src="https://user-images.githubusercontent.com/720814/71544576-8cebb900-2981-11ea-8376-cccb9eaf329f.png">

Requesting Camera Permiion
<img width="423" alt="image" src="https://user-images.githubusercontent.com/720814/71544589-b573b300-2981-11ea-906f-061e5253497f.png">

THis is also correctly shown in Seciroty and Privacy now.

In addition I cleaned the makefile..

sthalik commented:
> Perhaps I am confused between what opentrack-hier-str/opentrack-hier-pfx means. What do these acronym mean?

They aren't well thought out. I added them in a hurry, then they stayed that way.

> I do realise that not all plugins compile yet, but perhaps overtime we can see where the needs are.

Is it any worse than it originally was?

The diff looks good overall from visual inspection (some thunks should be removed, as per comments). I don't like the changes to `Find*.cmake` scripts. The originals should work as-is.

rvt commented:
> > Perhaps I am confused between what opentrack-hier-str/opentrack-hier-pfx means. What do these acronym mean?
> 
> They aren't well thought out. I added them in a hurry, then they stayed that way.

Fair enough.... been there done that myself :)
> 
> > I do realise that not all plugins compile yet, but perhaps overtime we can see where the needs are.
> 
> Is it any worse than it originally was?
Fair point... I am good to roll with it.

> 
> The diff looks good overall from visual inspection (some thunks should be removed, as per comments). I don't like the changes to `Find*.cmake` scripts. The originals should work as-is.

Can you give me an example of an change in the cmakefile you don't like?


Note: xplane plugin on XP11 also works again on OSX with latest unstable merged.

sthalik commented:
> Can you give me an example of an change in the cmakefile you don't like?

- `cmake/opentrack-install.cmake` breaks Wine output method
- `cmake/opentrack-platform.cmake` copy-paste error
- `cmake/opentrack-qt.cmake` can't install these for win32
- `pose-widget/CMakeLists.txt` debug code

Otherwise, I don't see problems that could break win32 or Linux.

sthalik commented:
Oh and add yourself to authors section in `README.md`.

sthalik commented:
Are your changes ready for merge? Also, would you like a commit bit?

rvt commented:
> > Are your changes ready for merge? Also, would you like a commit bit??
> 
> * `cmake/opentrack-install.cmake` breaks Wine output method
> * `cmake/opentrack-platform.cmake` copy-paste error
> * `cmake/opentrack-qt.cmake` can't install these for win32
> * `pose-widget/CMakeLists.txt` debug code
> 

I would like to take a look at these 4 files and draw a conclusion. I can do that today. After that I think it's fine.

There are still one issue but I think they can be handled in different commits, or at least be handled in the form of documentation:

* App-Nap, the application is put to sleep when running in background

rvt commented:
* `cmake/opentrack-install.cmake` : I am not sure why not including the DLL and the executable would break wine. As far as I can tell Wine is not used on OSX... But perhaps I am missing to understand a specific feature as I don't know TrackIR and it's tools..

*` cmake/opentrack-qt.cmake` : Should we have `Qt5::DBus` and `Qt5::Multimedia` to be OSX specific? I think there is not an issue adding these to windows if not used, to keep build files simpler... by preventing a other `#ifdef __APPLE__ `

*  `pose-widget/CMakeLists.txt`: Fixed

* `cmake/opentrack-platform.cmake` : Fixed

I f you are ok/agree with the above then it can be merged

sthalik commented:
> I am not sure why not including the DLL and the executable would break wine. As far as I can tell Wine is not used on OSX... But perhaps I am missing to understand a specific feature as I don't know TrackIR and it's tools..

The Wine protocol allows for Windows games to interface with opentrack. You can find `npclient.c` in `contrib/` for reference.

> Should we have `Qt5::DBus` and `Qt5::Multimedia` to be OSX specific? I think there is not an issue adding these to windows if not used, to keep build files simpler...

Build size. But this is debatable.

sthalik commented:
FWIW, see `contrib/trackir-client` on OSX with the Wine protocol.

rvt commented:
For my understanding: Wine is started on OSX to play a windows game. Then OpenTrack is started on OSX for head tracking towards that game?

sthalik commented:
> For my understanding: Wine is started on OSX to play a windows game. Then OpenTrack is started on OSX for head tracking towards that game?

> I will ad them in a separate deployment directory....

The path needs to match since the Wine plugin (along with the winegcc-treated executable) sets the Wine registry entry. They need to be installed in the libexec directory as-is.

rvt commented:
added thirdparty libs on generated DMG.

proto-wine does not build yet when SDK_WINE is enabled. I would suggest we will wait for users so we can also test this when fixed? I don't have anything to test on.

sthalik commented:
> added thirdparty libs on generated DMG.
> proto-wine does not build yet when SDK_WINE is enabled.

There's no point testing if it copies to the wrong directory. This is the most often used output method since OSX/Linux games are few and far in between.

What's the benefit of making a `.dmg` package? Excuse my OSX ignorance.

rvt commented:
> > added thirdparty libs on generated DMG.
> > proto-wine does not build yet when SDK_WINE is enabled.
> 
> There's no point testing if it copies to the wrong directory. This is the most often used output method since OSX/Linux games are few and far in between.

I am at a loss here. Never used wine or played any games running on wine... I was expected the user to know what to do with the files because DLL and EXE don't belong next to opentrack....

Building currently fails because of the wineg++ wrappers missing.

```
/opt/source/opentrack/build/proto-wine && wineg++ -mconsole -g -DNOMINMAX -O2 -m32 -std=c++17 -fPIC -o opentrack-wrapper-wine.exe -I /opt/source/opentrack -I /opt/source/opentrack/build /opt/source/opentrack/proto-wine/opentrack-wrapper-wine-main.cxx /opt/source/opentrack/proto-wine/opentrack-wrapper-wine-posix.cxx /opt/source/opentrack/proto-wine/opentrack-wrapper-wine-windows.cxx -Wall -Wextra -Wpedantic
/bin/sh: wineg++: command not found
```

> 
> What's the benefit of making a `.dmg` package? Excuse my OSX ignorance.

a DMG file can be used for distribution of an application. It's like a zip file but for OSX. You can than drag/drop the application in place, usually in the applications folder (see the drag pointer to the right). So we can upload the DMG file as a release and we are good. 
the DMG also follows a specific structure and includes all required libraries. Unlike Windows no DLL hell with DLLs that go into specific directories....

sthalik commented:
> I am at a loss here. Never used wine or played any games running on wine... I was expected the user to know what to do with the files because DLL and EXE don't belong next to opentrack....

I'd be hard-pressed to find any sim games running on non-Windows platforms. Take for instance Falcon BMS, DCS, Arma, and the like. These Windows-specific files belong with opentrack specifically for that reason.

> Building currently fails because of the wineg++ wrappers missing.

They shouldn't be missing on the build machine.

Does your PR introduce regressions for more specific reasons? It's fine as long as things that worked before, work still.

> the DMG also follows a specific structure and includes all required libraries. Unlike Windows no DLL hell with DLLs that go into specific directories....

That wasn't the case with the original structure produced by `make install`, as long as non-framework Qt is used.

rvt commented:
> > I am at a loss here. Never used wine or played any games running on wine... I was expected the user to know what to do with the files because DLL and EXE don't belong next to opentrack....
> 
> I'd be hard-pressed to find any sim games running on non-Windows platforms. Take for instance Falcon BMS, DCS, Arma, and the like. These Windows-specific files belong with opentrack specifically for that reason.
> 
> > Building currently fails because of the wineg++ wrappers missing.
> 
> They shouldn't be missing on the build machine.
> 
> Does your PR introduce regressions for more specific reasons? It's fine as long as things that worked before, work still.
> 
> > the DMG also follows a specific structure and includes all required libraries. Unlike Windows no DLL hell with DLLs that go into specific directories....
> 
> That wasn't the case with the original structure produced by `make install`, as long as non-framework Qt is used.

I think it would be nice that next to the Windows Installer we now also have a OSX downloadable version. This will save people the hassle from building OpenTrack. I think that would be difficult or impossible for all non tech people.

I would like to propose for this initial Merge Request not to include proto-wine. I will be looking into building it as I think it would be a very nice to have but for the first MVP we can see if this will atract people, see what is happening?

rvt commented:
Quick update. Installing wine from source was easy and that brings in winegcc, however build does fail:

```
winebuild -v -munix -fno-asynchronous-unwind-tables --cc-cmd=gcc -m64 --ld-cmd=ld -m32 -D_REENTRANT -fPIC --exe -o opentrack-wrapper-wine.exe-SqCRv6.spec.o -F opentrack-wrapper-wine.exe --subsystem console -L/Users/rvt/wine/lib64/wine -L/Users/rvt/wine/lib64 -- opentrack-wrapper-wine-main-nglUCd.o opentrack-wrapper-wine-posix-qHtqcE.o opentrack-wrapper-wine-windows-PJJVab.o /Users/rvt/wine/lib64/wine/libadvapi32.def /Users/rvt/wine/lib64/wine/libuser32.def /Users/rvt/wine/lib64/wine/libwinecrt0.a /Users/rvt/wine/lib64/wine/libkernel32.def /Users/rvt/wine/lib64/wine/libntdll.def 
winegcc: winebuild failed
```

Some of the usual tricks I tried:
 add -v ,, show all commands executed and run each command seperate. When I do that I don't get any error but I don't end up with an `opentrack-wrapper-wine.exe.so`. Having no experience with wine or winegcc and winebuild I am actually at a loss what else I could try, a bit of input would be nice.

sthalik commented:
Does the moral equivalent of "hello world" compile with winegcc?
